core/modules/path/path.module

```diff
@@ -110,9 +154,282 @@ function path_menu() {
+  }
+  $items[$site_404 . '/add-redirect'] = array(
+    'title' => 'Add URL redirect from this page to another location',
+    'type' => MENU_LOCAL_ACTION,
+    'page callback' => 'redirect_add_redirect_page',
+    'access callback' => 'redirect_access',
+    'access arguments' => array('create', 'redirect'),
+    'file' => 'redirect.admin.inc',
+  );
+
+  // If the database logging module is enabled, add special 404 listing pages.
+  if (module_exists('dblog')) {
+    $items['admin/config/search/redirect/404'] = array(
+      'title' => 'Fix 404 pages',
+      'description' => 'Add redirects for 404 pages.',
+      'page callback' => 'redirect_404_list',
+      'access arguments' => array('administer redirects'),
+      'file' => 'redirect.admin.inc',
+      'type' => MENU_LOCAL_TASK,
+      'weight' => 20,
+    );
+    $items['admin/reports/page-not-found/redirect'] = array(
+      'title' => 'Fix 404 pages with URL redirects',
+      'page callback' => 'backdrop_goto',
+      'page arguments' => array('admin/config/search/redirect/404'),
+      'access arguments' => array('administer redirects'),
+      'type' => MENU_LOCAL_ACTION,
+    );
+  }
+
+  // Devel generate integration.
+  if (module_exists('devel_generate')) {
+    $items['admin/config/development/generate/redirects'] = array(
+      'title' => 'Generate redirects',
+      'description' => 'Generate a given number of redirects. Optionally delete current redirects.',
+      'page callback' => 'backdrop_get_form',
+      'page arguments' => array('redirect_generate_form'),
+      'access arguments' => array('administer redirects'),
+      'file' => 'redirect.generate.inc',
+    );
+    $items['admin/config/search/redirect/generate'] = $items['admin/config/development/generate/redirects'];
+    $items['admin/config/search/redirect/generate']['type'] = MENU_LOCAL_ACTION;
+  }
+
   return $items;
 }
 
+function redirect_set_current_redirect($redirect) {
+  $static = &backdrop_static(__FUNCTION__);
+  $static = $redirect;
+}
+
+function redirect_get_current_redirect() {
+  $redirect = backdrop_static('redirect_set_current_redirect', NULL);
+
+  // If a redirect has not been set with redirect_set_current_redirect(), then
+  // attempt to find a redirect matching the current path, query string, and
+  // language code.
+  if (!isset($redirect)) {
+    $redirect = redirect_load_by_source(current_path(), $GLOBALS['language']->langcode, backdrop_get_query_parameters());
+  }
+
+  // @todo Add an alter hook here?
+  return $redirect;
+}
+
+/**
+ * Implements hook_url_inbound_alter().
+ */
+function path_url_inbound_alter(&$path, $original_path, $path_language) {
+  // If the current path global does not exist, then backdrop_get_path_alias()
+  // will fail. This condition only happens when $path is the front page.
+  // @todo Remove when http://drupal.org/node/1329914 is fixed in core.
+  if (empty($_GET['q'])) {
+    $_GET['q'] = config_get('system.core', 'site_frontpage');
+    return;
+  }
```

Comment:
This has been fixed in Drupal (and Backdrop) for a while. We can remove this condition.